## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('client_dataset.csv')
df

/Users/samriddhimangal/anaconda3/lib/python3.11/site-packages/pandas/core/arrays/masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year         item_id     category  \
0              17        2023  EUD29711-63-6U        decor   
1              20        2023  XDA18116-89-4A  consumables   
2               4        2023  ABE59463-05-7E     software   
3              17        2023  ZMM00836-65-0C  consumables   
4              21        2023  BZX55559-12-3X  consumables   
...           ...         ...             ...          ...   
54634          17        2023  UCA28532-31-0B     software   
54635           6        2023  EBD29277-23-7U  consumables   
54636           9        2023  EBD30093-55-9Z  consumables   
54637           3        2023  XMZ38239-49-3D    furniture   
54638           1        2023  ZCZ05589-50-5D     software   

              subcategory  unit_price  unit_cost  unit_weight   qty  \
0                wall art     1096.80     762.71         7.50   105   
1                    pens       24.95      15.09         1.49    21   
2      project management       13.52       7.86         1.68    39   
3                    pens       36.42      24.85         1.23    29   
4                    misc      195.10     108.17        46.43    20   
...                   ...         ...        ...          ...   ...   
54634           logistics       83.13      51.60         2.25    33   
54635    kitchen supplies      206.59     175.46        11.70    47   
54636   bathroom supplies       65.66      57.31         4.16   475   
54637              stands        1.48       1.22        18.04   112   
54638  project management        3.01       1.61         2.07  1031   

       line_number  
0                1  
1                0  
2                6  
3                3  
4                1  
...            ...  
54634            8  
54635            0  
54636            6  
54637            0  
54638            5  

[54639 rows x 18 columns]

In [2]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
pd.options.display.float_format = '{:.2f}'.format
df.describe()

client_id   order_id  order_week  order_year  unit_price  unit_cost  \
count   54639.00   54639.00    54639.00    54639.00    54639.00   54639.00   
mean    54837.87 5470190.03       11.36     2022.99      136.27      99.45   
std     25487.44 2599806.93        7.02        0.08      183.87     133.16   
min     10033.00 1000886.00        1.00     2022.00        0.01       0.01   
25%     33593.00 3196372.00        6.00     2023.00       20.80      14.84   
50%     53305.00 5496966.00       11.00     2023.00       68.31      49.89   
75%     78498.00 7733869.00       17.00     2023.00      173.16     125.57   
max     99984.00 9998480.00       52.00     2023.00     1396.23     846.27   

       unit_weight        qty  line_number  
count     54639.00   54639.00     54639.00  
mean          5.00     570.26         2.98  
std           5.33   18795.52         2.44  
min           0.00       0.00         0.00  
25%           1.45      32.00         1.00  
50%           3.24      68.00         3.00  
75%           6.89     170.00         5.00  
max          46.43 3958244.00         9.00

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
# What three item categories had the most entries?
df.category.value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?

df_consumables=df[df['category']=='consumables']
     
df_consumables.subcategory.value_counts().head(1)

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
df_client=df['client_id'].value_counts()
top_5=df_client.head(5)
top_5

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top_5.reset_index()
#type(top_5)
top_clients_by_quantity=top_5.index.to_list()

top_clients_by_quantity


[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the client with the most entries order order?
df[df['client_id']==33615].qty.sum()

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal']=df['unit_price']*df['qty']
df[['unit_price','qty','line_subtotal']]

unit_price   qty  line_subtotal
0         1096.80   105      115164.00
1           24.95    21         523.95
2           13.52    39         527.28
3           36.42    29        1056.18
4          195.10    20        3902.00
...           ...   ...            ...
54634       83.13    33        2743.29
54635      206.59    47        9709.73
54636       65.66   475       31188.50
54637        1.48   112         165.76
54638        3.01  1031        3103.31

[54639 rows x 3 columns]

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['total_weight']=df['unit_weight']*df['qty']

# Define a function
def shipping_price(total_weight):
    if total_weight < 50:
        return total_weight * 10
    return total_weight * 7

# "Apply" the function to the shipping_price column



df['shipping_price']=df['total_weight'].apply(shipping_price)
df[['unit_price','unit_weight','qty','total_weight','shipping_price']]

unit_price  unit_weight   qty  total_weight  shipping_price
0         1096.80         7.50   105        787.50         5512.50
1           24.95         1.49    21         31.29          312.90
2           13.52         1.68    39         65.52          458.64
3           36.42         1.23    29         35.67          356.70
4          195.10        46.43    20        928.60         6500.20
...           ...          ...   ...           ...             ...
54634       83.13         2.25    33         74.25          519.75
54635      206.59        11.70    47        549.90         3849.30
54636       65.66         4.16   475       1976.00        13832.00
54637        1.48        18.04   112       2020.48        14143.36
54638        3.01         2.07  1031       2134.17        14939.19

[54639 rows x 5 columns]

In [13]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_price']=(df['line_subtotal']+df['shipping_price'])*(1+0.0925)
df[['line_subtotal','shipping_price','line_price']]

line_subtotal  shipping_price  line_price
0          115164.00         5512.50   131839.08
1             523.95          312.90      914.26
2             527.28          458.64     1077.12
3            1056.18          356.70     1543.57
4            3902.00         6500.20    11364.40
...              ...             ...         ...
54634        2743.29          519.75     3564.87
54635        9709.73         3849.30    14813.24
54636       31188.50        13832.00    49184.90
54637         165.76        14143.36    15632.71
54638        3103.31        14939.19    19711.43

[54639 rows x 3 columns]

In [14]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).

df['line_cost']=(df['unit_cost']* df['qty'])+df['shipping_price']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number  line_subtotal  total_weight  shipping_price  line_price  \
0                1      115164.00        787.50         5512.50   131839.08   
1                0         523.95         31.29          312.90      914.26   
2                6         527.28         65.52          458.64     1077.12   
3                3        1056.18         35.67          356.70     1543.57   
4                1        3902.00        928.60         6500.20    11364.40   
...            ...            ...           ...             ...         ...   
54634            8        2743.29         74.25          519.75     3564.87   
54635            0        9709.73        549.90         3849.30    14813.24   
54636            6       31188.50       1976.00        13832.00    49184.90   
54637            0         165.76       2020.48        14143.36    15632.71   
54638            5        3103.31       2134.17        14939.19    19711.43   

       line_cost  
0       85597.05  
1         629.79  
2         765.18  
3        1077.35  
4        8663.60  
...          ...  
54634    2222.55  
54635   12095.92  
54636   41054.25  
54637   14280.00  
54638   16599.10  

[54639 rows x 23 columns]

In [15]:
# Create a column for the profit of each line using line cost and line price
df['line_profit']=df['line_price']-df['line_cost']
df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_cost unit_weight   qty  line_number  \
0              17        2023  ...    762.71        7.50   105            1   
1              20        2023  ...     15.09        1.49    21            0   
2               4        2023  ...      7.86        1.68    39            6   
3              17        2023  ...     24.85        1.23    29            3   
4              21        2023  ...    108.17       46.43    20            1   
...           ...         ...  ...       ...         ...   ...          ...   
54634          17        2023  ...     51.60        2.25    33            8   
54635           6        2023  ...    175.46       11.70    47            0   
54636           9        2023  ...     57.31        4.16   475            6   
54637           3        2023  ...      1.22       18.04   112            0   
54638           1        2023  ...      1.61        2.07  1031            5   

       line_subtotal  total_weight  shipping_price  line_price  line_cost  \
0          115164.00        787.50         5512.50   131839.08   85597.05   
1             523.95         31.29          312.90      914.26     629.79   
2             527.28         65.52          458.64     1077.12     765.18   
3            1056.18         35.67          356.70     1543.57    1077.35   
4            3902.00        928.60         6500.20    11364.40    8663.60   
...              ...           ...             ...         ...        ...   
54634        2743.29         74.25          519.75     3564.87    2222.55   
54635        9709.73        549.90         3849.30    14813.24   12095.92   
54636       31188.50       1976.00        13832.00    49184.90   41054.25   
54637         165.76       2020.48        14143.36    15632.71   14280.00   
54638        3103.31       2134.17        14939.19    19711.43   16599.10   

       line_profit  
0         46242.03  
1           284.47  
2           311.94  
3           466.22  
4          2700.80  
...            ...  
54634      1342.32  
54635      2717.32  
54636      8130.65  
54637      1352.71  
54638      3112.33  

[54639 rows x 24 colum

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [16]:
# Check your work using the totals above
orderID=[ 2742071,2173913,6128929]
totalprice=round(df.groupby(['order_id'])['line_price'].sum(),2)

for order in orderID:
   print(f"Order ID {order} had a total price of ${totalprice[order]:,.2f}")
   


Order ID 2742071 had a total price of $152,811.90
Order ID 2173913 had a total price of $162,388.72
Order ID 6128929 had a total price of $923,441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [17]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.

Client_price=df.groupby(['client_id'])['line_price'].sum()

for clients in top_clients_by_quantity:
   print(f"Client ID {clients}: ${Client_price[clients]:,.2f}")


Client ID 33615: $8,377,144.70
Client ID 66037: $10,259,514.80
Client ID 46820: $9,743,794.32
Client ID 38378: $12,906,550.88
Client ID 24741: $82,268,891.98


In [18]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
pd.options.display.float_format = '{:,.2f}'.format
df_topClients=df[df['client_id'].isin(top_clients_by_quantity)]

df_topClients_sum= df_topClients.groupby(['client_id'])[['qty','shipping_price','line_price','line_cost','line_profit']].sum()
df_topClients_sum=df_topClients_sum.reset_index()
df_topClients_sum

client_id     qty  shipping_price    line_price     line_cost   line_profit
0      24741  239862    5,126,448.37 82,268,891.98 45,688,899.71 36,579,992.27
1      33615   64313    1,828,834.89  8,377,144.70  6,175,163.91  2,201,980.79
2      38378   73667    3,429,455.40 12,906,550.88  9,634,720.98  3,271,829.90
3      46820   75768    1,601,448.84  9,743,794.32  7,007,191.64  2,736,602.68
4      66037   43018    1,395,151.85 10,259,514.80  7,004,482.98  3,255,031.82

In [19]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 
col={'client_id':'Client ID',
     'qty':'Units',
    'shipping_price':'Shipping',
    'line_price':'Total Revenue',
     'line_cost':'Total Cost',
     'line_profit':'Total Profit' 
}
df_topClients_sum=df_topClients_sum.rename(columns=col)

In [20]:
# Define a function that converts a dollar amount to millions.

def currency_format_millions(amount):
    return (amount/1000000)

In [21]:
# Apply the currency_format_millions function to only the money columns.


# To ensure that original dataframe is returned if program is re-run a  reset function is used  that returns the original DataFrame
def reset_dataframe(original_df):
    return original_df.copy()

# Create a copy of the original DataFrame for resetting
original_df_topClients_sum = df_topClients_sum.copy()

# Reset the DataFrame before applying the conversion
df_topClients_sum_mil = reset_dataframe(original_df_topClients_sum)


df_topClients_sum_mil['Shipping']=df_topClients_sum_mil['Shipping'].apply(currency_format_millions)
df_topClients_sum_mil['Total Revenue']=df_topClients_sum_mil['Total Revenue'].apply(currency_format_millions)
df_topClients_sum_mil['Total Cost']=df_topClients_sum_mil['Total Cost'].apply(currency_format_millions)
df_topClients_sum_mil['Total Profit']=df_topClients_sum_mil['Total Profit'].apply(currency_format_millions)



In [22]:

# Rename the columns to reflect the change in the money format. 

col_mil={'Shipping':'Shipping (millions)',
    'Total Revenue': 'Total Revenue (millions)',
     'Total Cost': 'Total Cost(millions) ',
    'Total Profit' : 'Total Profit (millions)'
}
df_topClients_summary_millions=df_topClients_sum_mil.rename(columns=col_mil)

df_topClients_summary_millions

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862                 5.13                     82.27   
1      33615   64313                 1.83                      8.38   
2      38378   73667                 3.43                     12.91   
3      46820   75768                 1.60                      9.74   
4      66037   43018                 1.40                     10.26   

   Total Cost(millions)   Total Profit (millions)  
0                  45.69                    36.58  
1                   6.18                     2.20  
2                   9.63                     3.27  
3                   7.01                     2.74  
4                   7.00                     3.26

In [23]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
df_topClients_summary_millions_sort=df_topClients_summary_millions.sort_values(by="Total Profit (millions)", ascending=False)
df_topClients_summary_millions_sort

Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
0      24741  239862                 5.13                     82.27   
2      38378   73667                 3.43                     12.91   
4      66037   43018                 1.40                     10.26   
3      46820   75768                 1.60                      9.74   
1      33615   64313                 1.83                      8.38   

   Total Cost(millions)   Total Profit (millions)  
0                  45.69                    36.58  
2                   9.63                     3.27  
4                   7.00                     3.26  
3                   7.01                     2.74  
1                   6.18                     2.20

# Summary

**Background:**

The above wholesale dataset consists of 54,639 sales records and 18 features for the year 2023 (order_date feature).This dataset was used to identify top five clients(customers) based on how frequently they have visited or ordered from the store. These clients were further analyzed and following observations were made.

 * Revenue: These top five clients bring around $123.56 million revenue in total for the store and out of which        66.6 percent of revenue comes from single client with id 24741.
 
*  Profit: Also, it may be noted that store earns around $48.05 million profit from these top 5 clients out of        which 76.1% is from the single client with id 24741.

*  Units Sold: A total of 496,628 units were sold to these top 5 clients, out of which the clientID:24741 accounts    for 48.3%.

**Further Analysis Suggestions:**

* Revenue/Profit based Client Ranking: Top clients based on total revenue and total profit can be identified.

* Product Segments: Segments of products that bring most revenue and are most popular among clients may be           identified.

* Seasonal Trends: As the data is for the year 2023 based on order_date, seasonal trends of product sold can be       determined.

**Utility of Insights:**

These insights will help in:

* Customizing promotions for specific clients.

* Targeting marketing strategies based on product popularity and seasonality.

* Enhancing overall sales and customer engagement strategies.
